### DISCUSSION TITLE
Does it support dynamic model generation?

### DISCUSSION THREAD

**Discussion by berislavlopac** [koxudaxi/datamodel-code-generator#1160] [General]
Looking at the documentation, this project generates Python code as text, which can be saved into output files to be used/imported as modules.

With Python's dynamic capabilities, it is quite easy to -- instead of textual code -- generate actual model classes on the fly. Is that supported as a use case at the moment?

**Comment by koxudaxi** [comment]
@berislavlopac 
the short answer is No.
Do you have a concrete idea to realize model generation on the fly?

**Reply by SebastianDahlin** [reply]
I have a use case:
1. Load JSON to python model containing both the class definitions and the data that should instaciate the classes.
2. I would like to, on the fly, create the models. All nested under a RootModel.
3. Parse the data using the RootModel. Ie. RootModel(**data).

Reason: I have an app I built in FastApi but can't get my colleagues to learn and use python :(
Instead I built an admin web interface that I am serving using FastApi, where a user can create classes and data with an intuitive interface. The created data is then saved as JSON on disk. 
The "real" app then uses the JSON on disk when it's run. 

Sorry for the long rant. Been looking for a solution for a long time and it feels like this could be a viable solution.

**Reply by berislavlopac** [reply]
> Do you have a concrete idea to realize model generation on the fly?

That depends on how does the code generation happen. In [momoa](https://github.com/berislavlopac/momoa) I have used [statham](https://github.com/jacksmith15/statham-schema)'s code generation to extract their intermediate model and expose it as a Python class in its own right; if DMCG contructs the class before converting it to text, the same approach could be used here.

If it doesn't, the only thing I can think of (and will test when I find some time) is to use `eval` (with all its caveats).

**Reply by berislavlopac** [reply]
@koxudaxi Probably the most promising option would be to use Pydantic's `create_model` function - https://docs.pydantic.dev/usage/models/#dynamic-model-creation

**Reply by koxudaxi** [reply]
Looks good.
I'm not familiar with `create_model`.
We should confirm the different specs like module level.
Also, I don't have the time. PR welcome.

**Reply by berislavlopac** [reply]
Yeah, I hear you with the lack of time. I'm currently exploring the viable options, and have a few thoughts so far:

1. The `parser.parse` internally builds a collection of models, but they can't be directly accessed; we will need either to extract that into a separate function, or to add an optional argument to `parse` to make it return the collection.
2. I haven't yet figured out how those models are converted to Pydantic ones, or dataclass; my idea is to use `create_model` to do that conversion. Not sure yet how it would work with dataclasses exactly, but just getting it to work with Pydantic would be a great starting point.

**Reply by koxudaxi** [reply]
@berislavlopac 

> 1.
`parse()` returns model as `str`. we change the return value or refactoring and add a new method to get a dynamic model.
https://github.com/koxudaxi/datamodel-code-generator/blob/3c6958275a7a2793b4f91b1709660fd86779b88c/datamodel_code_generator/parser/base.py#L955-L960

> 2.
After parsing, the models are created.
 The CLI calls `__str__` method on the model to render the Jinja template.
https://github.com/koxudaxi/datamodel-code-generator/blob/3c6958275a7a2793b4f91b1709660fd86779b88c/datamodel_code_generator/model/base.py#L196-L204

`BaseModel` is here.
https://github.com/koxudaxi/datamodel-code-generator/blob/3c6958275a7a2793b4f91b1709660fd86779b88c/datamodel_code_generator/model/pydantic/base_model.py#L187

`Dataclass`
https://github.com/koxudaxi/datamodel-code-generator/blob/3c6958275a7a2793b4f91b1709660fd86779b88c/datamodel_code_generator/model/dataclass.py#L15

I think we can add `create_model` method on `DataModel`.
`DataModel` has all context to create mode dynamically. https://github.com/koxudaxi/datamodel-code-generator/blob/master/datamodel_code_generator/model/template/pydantic/BaseModel.jinja2

I need to learn how to treat modulable models.
Can we set the namespace(module name)  to a dynamic model? :thinking:

**Reply by berislavlopac** [reply]
1. Yes, that's exactly what I meant.

For 2:
> I think we can add create_model method on DataModel.

Maybe better on its subclasses, as I expect the mechanism to be different for Pydantic and dataclasses.

> `DataModel` has all context to create mode dynamically

Yes, that was my conclusion.

> Can we set the namespace(module name) to a dynamic model?

For dynamically created classes, the module name is `__main__`:
```python
>>> Foo = type("Foo", (), {})
>>> Foo.__module__
'__main__'
```

We can probably also use the [types](https://docs.python.org/3/library/types.html) library to define namespaces. 🤔

**Comment by Danipulok** [comment]
@berislavlopac good day. Have you created a PR or something similar in this direction?

**Reply by berislavlopac** [reply]
I'm afraid not, been too swamped with work. 😞 But it's been on my mind and I keep planning to return to it.

**Reply by Danipulok** [reply]
Got it, thanks a for a quick reply!
Good luck

**Comment by koxudaxi** [comment]
Thank you for the feature request — this has now been released!

**Comment by Danipulok** [comment]
@koxudaxi hey! I think I vaguely remember that I was searching for JsonSchema -> Pydantic models. So does it mean it's supported now? If yes, could you please link related docs/pr/issue? Thank you very much for implementing this feature!

**Reply by koxudaxi** [reply]
Thank you for remembering the CLI.
I created a page for the feature!
https://datamodel-code-generator.koxudaxi.dev/dynamic-model-generation/

**Reply by berislavlopac** [reply]
@koxudaxi First of all, thank you for the great addition, I've been wanting this feature forever! :heart: 

That being said, I'm having some issues with [custom configuration](https://datamodel-code-generator.koxudaxi.dev/dynamic-model-generation/#with-custom-configuration).

On `from datamodel_code_generator import GenerateConfig` I'm getting the `ImportError`:

```
ImportError: cannot import name 'GenerateConfig' from 'datamodel_code_generator'
```

I was able to import the components individually, but that requires rebuilding the `GenerateConfig` class:

```python
from datamodel_code_generator.config import GenerateConfig
from datamodel_code_generator.dynamic import generate_dynamic_models
from datamodel_code_generator.enums import UnionMode

GenerateConfig.model_rebuild()
```

It looks like `GenerateConfig` is imported in your `__init__.py` only on `TYPE_CHECKING`. :thinking:
